ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java

```diff
@@ -0,0 +1,596 @@
+import java.util.concurrent.atomic.AtomicReference;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Test;
+
+
+/**
+ * Tests for {@link GetBlobOperation}
+ */
+public class GetBlobOperationTest {
+  private static final int MAX_PORTS_PLAIN_TEXT = 3;
+  private static final int MAX_PORTS_SSL = 3;
+  private static final int CHECKOUT_TIMEOUT_MS = 1000;
+  private static final int BLOB_USER_METADATA_SIZE = 10;
+
+  private final MockTime time = new MockTime();
+  private final Map<Integer, GetOperation> correlationIdToGetOperation = new HashMap<>();
+  private final Random random = new Random();
+  private final FutureResult<ReadableStreamChannel> operationFuture = new FutureResult<>();
+
+  private int requestParallelism = 2;
+  private int successTarget = 1;
+  private int maxChunkSize;
+  private int blobSize = 1000;
+  private RouterConfig routerConfig;
+  private MockClusterMap mockClusterMap;
+  private NonBlockingRouterMetrics routerMetrics;
+  private MockServerLayout mockServerLayout;
+  private int replicasCount;
+  private AtomicReference<MockSelectorState> mockSelectorState = new AtomicReference<>();
+  private ResponseHandler responseHandler;
+  private MockNetworkClientFactory networkClientFactory;
+  private NetworkClient networkClient;
+  private VerifiableProperties vprops;
+  private NonBlockingRouter router;
+  private String blobIdStr;
+  private BlobProperties blobProperties;
+  private byte[] userMetadata;
+  private byte[] putContent;
+
+  private final GetTestRequestRegistrationCallbackImpl requestRegistrationCallback =
+      new GetTestRequestRegistrationCallbackImpl();
+
+  private class GetTestRequestRegistrationCallbackImpl implements RequestRegistrationCallback<GetOperation> {
+    private List<RequestInfo> requestListToFill;
+
+    @Override
+    public void registerRequestToSend(GetOperation getOperation, RequestInfo requestInfo) {
+      requestListToFill.add(requestInfo);
+      correlationIdToGetOperation.put(((RequestOrResponse) requestInfo.getRequest()).getCorrelationId(), getOperation);
+    }
+  }
+
+  private final AtomicInteger operationsCount = new AtomicInteger(0);
+  private final OperationCompleteCallback operationCompleteCallback = new OperationCompleteCallback(operationsCount);
+
+  public GetBlobOperationTest() {
+    // Defaults. Tests should override these as appropriate.
+    maxChunkSize = random.nextInt(1024 * 1024) + 1;
+    blobSize = maxChunkSize * random.nextInt(10) + random.nextInt(maxChunkSize - 1) + 1;
+  }
+
+  @After
+  public void after() {
+    if (networkClient != null) {
+      networkClient.close();
+    }
+    Assert.assertEquals("All operations should have completed", 0, operationsCount.get());
+  }
+
+  /**
+   * Instantiate a router, perform a put, close the router. The blob that was put will be saved in the MockServer,
+   * and can be queried by the getBlob operations in the test.
+   */
+  private void doPut()
```

Comment:
would it be possible to order it such that all tests (public functions) are on the top and all helpers (private functions), at the bottom?